src/libpgagroal/utils.c

```diff
@@ -954,128 +955,124 @@ pgagroal_backtrace(void)
 
 #endif
 
+/* Parser for pgagroal-cli commands */
+
 bool
-parse_command(int argc,
-              char** argv,
-              int offset,
-              char* command,
-              char* subcommand,
-              char** key,
-              char* default_key,
-              char** value,
-              char* default_value)
+parse_command(int argc, char** argv, int offset, struct pgagroal_parsed_command* cmd,
+              const struct pgagroal_command command_table[], int command_count)
 {
+#define CONSUME(_x) _x = (offset < argc) ? argv[offset++] : NULL
+#define EMPTY_STR(_s) (_s[0] == 0)
+
+   char* command = NULL;
+   char* subcommand = NULL;
+   int command_match[MISC_LENGTH] = {COMMAND_UNKNOWN};
+   int command_match_count = 0;
 
-   // sanity check: if no arguments, nothing to parse!
+   /* Exit if nothing to parse */
    if (argc <= offset)
    {
+      sprintf(cmd->error_message, "A command is required\n");
       return false;
    }
 
-   // first of all check if the command is the same
-   // as the first argument on the command line
-   if (strncmp(argv[offset], command, MISC_LENGTH))
+   /* Parse command, and exit if there is no match */
+   CONSUME(command);
+   for (int i = 0; i < command_count; i++)
    {
+      if (strncmp(command, command_table[i].command, MISC_LENGTH) == 0)
+      {
+         command_match[command_match_count++] = i;
+      }
+   }
+   if (command_match_count == 0)
+   {
+      sprintf(cmd->error_message, "Invalid command '%s'\n", command);
       return false;
    }
 
-   if (subcommand)
+   /* Parse subcommand, and exit if there is no match */
+   CONSUME(subcommand);
+   subcommand = subcommand ? subcommand : "";
+   for (int i = 0; i < command_match_count; i++)
    {
-      // thre must be a subcommand check
-      offset++;
-
-      if (argc <= offset)
-      {
-         // not enough command args!
-         return false;
-      }
-
-      if (strncmp(argv[offset], subcommand, MISC_LENGTH))
+      if (strncmp(subcommand, command_table[command_match[i]].subcommand, MISC_LENGTH) == 0)
       {
-         return false;
+         cmd->command_code = command_match[i]; /* pushes down the matches */
       }
    }
-
-   if (key)
+   if (cmd->command_code == COMMAND_UNKNOWN)
    {
-      // need to evaluate the database or server or configuration key
-      offset++;
-      *key = argc > offset ? argv[offset] : default_key;
-      if (*key == NULL || strlen(*key) == 0)
+      if (EMPTY_STR(subcommand))
       {
-         goto error;
+         sprintf(cmd->error_message, "Command '%s' requires a subcommand\n", command);
+         return false;
       }
-
-      // do I need also a value?
-      if (value)
+      else
       {
-         offset++;
-         *value = argc > offset ? argv[offset] : default_value;
-
-         if (*value == NULL || strlen(*value) == 0)
+         offset--;
+         subcommand = "";
```

Comment:
I don't understand why `subcommand` is cleared here, since it could have been `CONSUME`d.
In fact:

```
% pgagroal-admin user foo
Invalid subcommand '' for command 'user'


% pgagroal-cli conf foo
Invalid subcommand '' for command 'conf'
```

Reply:
My mistake. It should be cleared at the end of this if. To be honest, the code is a little awkward here, I can improve it.